[node-addon-api] Add new port

Pospelove opened the pull request:
depends on https://github.com/microsoft/vcpkg/pull/26768

node-api is official C API for NodeJS

node-addon-api is C++ wrapper around node-api

Pospelove committed:
port is building for x64-windows

Pospelove committed:
Add file endings

Pospelove committed:
Reformat manifest file

Pospelove committed:
Add versions

Pospelove committed:
Remove comment

Pospelove committed:
Check host system instead of target

Pospelove committed:
fix user fs pollution & allow nodejs tool to have the same triplet as target has

Pospelove committed:
Merge branch 'node' of https://github.com/Pospelove/vcpkg into node

Pospelove committed:
x86-windows now builds

Pospelove committed:
Support find_package (TARGET_SOURCES still todo)

Pospelove committed:
add win_delay_hook.cc

Pospelove committed:
license for nodejs

Pospelove committed:
reformat

Pospelove committed:
reformat

Pospelove committed:
Update node-api.json

Pospelove committed:
nodejs port ver increment

Pospelove committed:
fix versions

Pospelove committed:
fix versions

Pospelove committed:
Fix find_library/find_path review comments

Pospelove committed:
node-api-config.cmake => unofficial-node-api-config.cmake

Pospelove committed:
remove vcpkg_execute_npm_command.cmake

Pospelove committed:
update git-tree in node-api.json

Pospelove committed:
fix node triplet

Pospelove committed:
remove obvious comment

Pospelove committed:
Update node-api.json

Pospelove committed:
Revert everything related to vcpkg-tool-nodejs

Pospelove committed:
fix port-version of vcpkg-tool-nodejs in baseline.json

Pospelove committed:
Update node-api.json

Pospelove committed:
ban arm

Pospelove committed:
fix EOF

Pospelove committed:
Update node-api.json

Pospelove committed:
Tweaks for Linux failure debugging

Pospelove committed:
Update node-api.json

Pospelove committed:
Reformat with vscode cmake formatter

Pospelove committed:
Even more debug logging

Pospelove committed:
add <node_root>/bin to search

Pospelove committed:
Update node-api.json

Pospelove committed:
Add debugging for 2-nd error encountered

Pospelove committed:
update ver

Pospelove committed:
update

Pospelove committed:
update ver

Pospelove committed:
add nodejs tool to path

Pospelove committed:
update ver

Pospelove committed:
use prefix npm option

Pospelove committed:
add version

Pospelove committed:
move things to node tool dir (temporary)

Pospelove committed:
add ver

Pospelove committed:
Add CMAKE_JS_LIB emptyness check

Pospelove committed:
add version

Pospelove committed:
separate node root ad node bin directories

Pospelove committed:
update node-api-config.cmake

Pospelove committed:
add ver

Pospelove committed:
typo-fi

Pospelove committed:
update version

Pospelove committed:
remove debug logging

Pospelove committed:
update ver

Pospelove committed:
do not use system cmake

Pospelove committed:
add ver

Pospelove committed:
rename vars

Pospelove committed:
Update ver

Pospelove committed:
Add comment

Pospelove committed:
add-version

Pospelove committed:
Now downloads cmake-js from github, but deps aren't downloaded (wip)

Pospelove committed:
add version

Pospelove committed:
Seems to work except lodash (wip)

Pospelove committed:
add version

Pospelove committed:
license null (wip)

Pospelove committed:
license null -reformat- (wip)

Pospelove committed:
add ver

Pospelove committed:
Add todo (wip)

Pospelove committed:
add ver

Pospelove committed:
install lodash, new error now related to yargs isn't being built (wip)

Pospelove committed:
add version (wip)

Pospelove committed:
fix (wip)

Pospelove committed:
add version (wip)

Pospelove committed:
Cleanup, will try one thing (wip)

Pospelove committed:
no cmake-js ver, now install includes (wip)

Pospelove committed:
⌛ closing temporary, I'll get back to this later ⌛. at least I know what to do

Pospelove committed:
test

Pospelove committed:
baseline

Pospelove committed:
format

Pospelove committed:
fill with zeros

Pospelove committed:
update supports

Pospelove committed:
correct url for win

Pospelove committed:
addver

Pospelove committed:
better handle arch and os

Pospelove committed:
ver

Pospelove committed:
update

Pospelove committed:
update

Pospelove committed:
add-todo

Pospelove committed:
add win support

Pospelove committed:
ver

Pospelove committed:
qol

Pospelove committed:
ver

Pospelove committed:
add-sha

Pospelove committed:
ver

Pospelove committed:
update

Pospelove committed:
update

Pospelove committed:
fix-license

Pospelove committed:
fix-license

Pospelove committed:
add sha

Pospelove committed:
add sha

Pospelove committed:
linux-arm64

Pospelove committed:
linux-arm64

Pospelove committed:
sha512 for arm

Pospelove committed:
sha512 for arm

Pospelove committed:
remove imported sources

Pospelove committed:
remove imported sources

Pospelove committed:
apply suggestion

Pospelove committed:
add ver

Pospelove committed:
fix eof

Pospelove committed:
add ver

Pospelove committed:
[node-addon-api] Add new port

Pospelove committed:
add version

Pospelove committed:
add newline

Pospelove committed:
add version

JonLiu1993 commented:
@Pospelove, Thanks for your pr, if this pr is ready for review please let me know.

Pospelove commented:
@JonLiu1993 thank you, it's ready

however merging this would depend on https://github.com/microsoft/vcpkg/pull/26768

Pospelove committed:
merge

Pospelove commented:
@JonLiu1993 ready

JonLiu1993 commented on ports/node-addon-api/portfile.cmake:
```diff
@@ -0,0 +1,14 @@
+  OUT_SOURCE_PATH SOURCE_PATH
+  REPO nodejs/node-addon-api
+  REF v5.0.0
+  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
```
```suggestion
  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
HEAD_REF main
```

JonLiu1993 commented on ports/node-addon-api/portfile.cmake:
```diff
@@ -0,0 +1,14 @@
+  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
+)
+
+file(COPY ${SOURCE_PATH}/napi.h DESTINATION ${CURRENT_PACKAGES_DIR}/include)
+file(COPY ${SOURCE_PATH}/napi-inl.h DESTINATION ${CURRENT_PACKAGES_DIR}/include)
+file(COPY ${SOURCE_PATH}/napi-inl.deprecated.h DESTINATION ${CURRENT_PACKAGES_DIR}/include)
```
```suggestion
file(COPY "${SOURCE_PATH}/napi.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
file(COPY "${SOURCE_PATH}/napi-inl.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
file(COPY "${SOURCE_PATH}/napi-inl.deprecated.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
```

JonLiu1993 commented on ports/node-addon-api/portfile.cmake:
```diff
@@ -0,0 +1,14 @@
+
+file(INSTALL ${SOURCE_PATH}/LICENSE.md DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
+
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/unofficial-node-addon-api-config.cmake DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT})
```
```suggestion
file(COPY "${CMAKE_CURRENT_LIST_DIR}/unofficial-node-addon-api-config.cmake" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
```

JonLiu1993 commented on ports/node-addon-api/vcpkg.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "node-addon-api",
+  "version-semver": "5.0.0",
```
```suggestion
  "version": "5.0.0",
```

Pospelove committed:
Update ports/node-addon-api/vcpkg.json

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

Pospelove committed:
Update ports/node-addon-api/portfile.cmake

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

Pospelove committed:
Update ports/node-addon-api/portfile.cmake

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

Pospelove committed:
Update ports/node-addon-api/portfile.cmake

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

JonLiu1993 commented on ports/node-addon-api/portfile.cmake:
```diff
@@ -0,0 +1,15 @@
+  OUT_SOURCE_PATH SOURCE_PATH
+  REPO nodejs/node-addon-api
+  REF v5.0.0
+  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
+HEAD_REF main
```
```suggestion
  SHA512 5ee14bc325ce66c6f645c3b638efaf69a7dfc90f0d6514f8568dedf34d5b5460b80c4438c4c12264d80c827edf46a82f0ecceb53b2a370ed5ec2ee640fa1ef7f
  HEAD_REF main
```

Pospelove committed:
Update ports/node-addon-api/portfile.cmake

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

Pospelove committed:
Update node-addon-api.json

JonLiu1993 commented:
@Pospelove , Thanks for your pr, when I tested the usage I got this error, please take a look:

CMakeLists.txt:
````
# CMakeList.txt : CMake project for CMakeProject42, include source and define
# project specific logic here.
#
cmake_minimum_required (VERSION 3.8)

project(test)

find_package(node-addon-api CONFIG REQUIRED)

add_executable (main "CMakeProject42.cpp" "CMakeProject42.h")

target_link_libraries(main PRIVATE unofficial::node-addon-api::node-addon-api)

````


CMakeProject42.cpp:
````
// CMakeProject42.cpp : Defines the entry point for the application.
//

#include "CMakeProject42.h"
#include "napi.h"

using namespace std;

int main()
{
	cout << "Hello CMake." << endl;
	return 0;
}

````
CMakeProject42.h:
````
// CMakeProject42.h : Include file for standard system include files,
// or project specific include files.

#pragma once

#include <iostream>

// TODO: Reference additional headers your program requires here.

````

````
1> CMake generation started for configuration: 'x64-Debug'.
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 16 2019" -A x64  -DCMAKE_CONFIGURATION_TYPES:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\vzhli17\source\repos\CMakeProject42\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE=F:/Feature-test/node-addon-api/vcpkg/scripts/buildsystems/vcpkg.cmake "C:\Users\vzhli17\source\repos\CMakeProject42" 2>&1"
1> Working directory: C:\Users\test\source\repos\CMakeProject42\out\build\x64-Debug
1> [CMake] -- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19045.
1> [CMake] -- The C compiler identification is MSVC 19.29.30146.0
1> [CMake] -- The CXX compiler identification is MSVC 19.29.30146.0
1> [CMake] -- Detecting C compiler ABI info
1> [CMake] -- Detecting C compiler ABI info - done
1> [CMake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting C compile features
1> [CMake] -- Detecting C compile features - done
1> [CMake] -- Detecting CXX compiler ABI info
1> [CMake] -- Detecting CXX compiler ABI info - done
1> [CMake] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting CXX compile features
1> [CMake] -- Detecting CXX compile features - done
1> [CMake] CMake Error at F:/Feature-test/node-addon-api/vcpkg/scripts/buildsystems/vcpkg.cmake:843 (_find_package):
1> [CMake]   Could not find a package configuration file provided by "node-addon-api"
1> [CMake]   with any of the following names:
1> [CMake] 
1> [CMake]     node-addon-apiConfig.cmake
1> [CMake]     node-addon-api-config.cmake
1> [CMake] 
1> [CMake]   Add the installation prefix of "node-addon-api" to CMAKE_PREFIX_PATH or set
1> [CMake]   "node-addon-api_DIR" to a directory containing one of the above files.  If
1> [CMake]   "node-addon-api" provides a separate development package or SDK, be sure it
1> [CMake]   has been installed.
1> [CMake] Call Stack (most recent call first):
1> [CMake]   CMakeProject42/CMakeLists.txt:10 (find_package)
1> [CMake] -- Configuring incomplete, errors occurred!
1> [CMake] See also "C:/Users/test/source/repos/CMakeProject42/out/build/x64-Debug/CMakeFiles/CMakeOutput.log".
1> 'C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 16 2019" -A x64  -DCMAKE_CONFIGURATION_TYPES:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject42\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE=F:/Feature-test/node-addon-api/vcpkg/scripts/buildsystems/vcpkg.cmake "C:\Users\test\source\repos\CMakeProject42" 2>&1"' execution failed with error: ''C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 16 2019" -A x64  -DCMAKE_CONFIGURATION_TYPES:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject42\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE=F:/Feature-test/node-addon-api/vcpkg/scripts/buildsystems/vcpkg.cmake "C:\Users\test\source\repos\CMakeProject42" 2>&1"' returned with exit code: 1'.


````

dg0yt commented:
> find_package(node-addon-api CONFIG REQUIRED)
> target_link_libraries(main PRIVATE unofficial::node-addon-api::node-addon-api)

Shouldn't this be
~~~cmake
find_package(unofficial-node-addon-api CONFIG REQUIRED)
~~~
or similar?

Pospelove commented:
wow. it seems both are not working for me for some reason. 

unofficial-node-addon-api and node-addon-api 

investigating

Pospelove commented:
I see. Fixing.

Pospelove committed:
Update portfile.cmake

Pospelove committed:
Update node-addon-api.json

dg0yt commented on ports/node-addon-api/unofficial-node-addon-api-config.cmake:
```diff
@@ -0,0 +1,13 @@
+find_package(unofficial-node-api CONFIG REQUIRED)
```
This should be 
~~~cmake
include(CMakeFindDependencyMacro)
find_dependency(unofficial-node-api CONFIG)
~~~

dg0yt commented on ports/node-addon-api/unofficial-node-addon-api-config.cmake:
```diff
@@ -0,0 +1,13 @@
+  NAMES napi.h
+  PATHS "${CMAKE_CURRENT_LIST_DIR}/../../include/node-addon-api"
+  NO_DEFAULT_PATH
+  REQUIRED)
```
I don't think `find_path` is needed here: The path is already known. Just insert into `INTERFACE_INCLUDE_DIRECTORIES`.

dg0yt commented on ports/node-addon-api/portfile.cmake:
```diff
@@ -0,0 +1,15 @@
+file(COPY "${SOURCE_PATH}/napi-inl.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
+file(COPY "${SOURCE_PATH}/napi-inl.deprecated.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
+
+file(INSTALL ${SOURCE_PATH}/LICENSE.md DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
Missing quotes.
I would prefer if this was the last line in the portfile.

Pospelove committed:
Update unofficial-node-addon-api-config.cmake

Pospelove committed:
Update node-addon-api.json

Pospelove commented:
@JonLiu1993 Works now. Changed your cmakelists a bit:

added cxx standard and "unofficial" prefix for find_package

```cmake
set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_LIST_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake")

# CMakeList.txt : CMake project for CMakeProject42, include source and define
# project specific logic here.
#
cmake_minimum_required (VERSION 3.8)

set(CMAKE_CXX_STANDARD 20) # !!! I added this 

project(test)

find_package(unofficial-node-addon-api CONFIG REQUIRED) # !!! and added unofficial word here

add_executable (main "CMakeProject42.cpp" "CMakeProject42.h")

target_link_libraries(main PRIVATE unofficial::node-addon-api::node-addon-api)
```

@dg0yt thanks. taking a look

Pospelove committed:
fix dg0yt's comments 1/

Pospelove committed:
fix 2/

Pospelove committed:
fix 3/3

Pospelove commented:
@dg0yt  @JonLiu1993  works and fixed

Pospelove commented:
@JonLiu1993

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into node-addon-api

JonLiu1993 commented:
> @JonLiu1993

@Pospelove, sorry for reply you so late, after the CI turns green, I will put the reviewed label on the table after the local test is successful.

Pospelove commented:
@JonLiu1993
plz consider taking a look

Pospelove commented:
@JonLiu1993

JonLiu1993 commented:
@Pospelove, sorry to reply you so late, we just came back from vacation, so we didn't respond to you in time.

JonLiu1993 committed:
update version

JonLiu1993 commented:
Tested usage by `node-addon-api:x64-windows` triplet successfully.

BillyONeal committed:
Add usage.

BillyONeal reviewed (approved):
Thanks for the new port!

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.  
  Note: There is a 5.1.0 already :)
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See docs/examples/adding-an-explicit-usage.md for context.  
  I added usage text
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.
